samples/nrf9160/http_application_update/src/main.c

```diff
@@ -115,6 +124,16 @@ void fota_dl_handler(enum fota_download_evt_id evt_id)
 		break;
 	case FOTA_DOWNLOAD_EVT_ERROR:
 		printk("Received error from fota_download\n");
+		if (restarts_left > 0) {
+			restarts_left--;
+			printk("Attempting restart (%d retries left)\n",
+				restarts_left);
+			/* We cannot call fota_download_restart until after
+			 * this callback handler has completed, so we use
+			 * k_delayed_work
+			 */
+			k_delayed_work_submit(&fota_work, 2);
```

Comment:
I'll put the K_MSEC(2) in...

But while we're here, this line bothers me.  What needs to happen is that the work item shouldn't be triggered until after the download_client thread has had a chance to hit the suspend at the top of its loop.  I couldn't think of any way to guarantee that.  I just wanted the shortest delay possible that would allow the handler to return.  But even then, in a bonkers setup, if the work queue thread has a higher priority than the download_client thread and an interrupt comes by in the wrong time window, this work queue item could *still* hit before the download client is suspended.

Is there another way to guarantee that the download_client has hit the suspend before we try to restart?

Reply:
As I hinted at when I opened this, if the download_client event handler paid attention to the return value it got back from DOWNLOAD_CLIENT_EVT_ERROR events and one of them could mean "the socket has been closed and reopened and your offsets/counters adjusted, just keep going" while the other would match the current "stop working", that would sidestep this whole thing very cleanly.

I didn't want to change the download_client API like that without more conversation, but maybe this is the time for that conversation?